src/prefect/new_task_engine.py

```diff
@@ -0,0 +1,382 @@
+            await self.set_state(Retrying(), force=True)
+            self.retries = self.retries + 1
+            return True
+        return False
+
+    async def handle_exception(self, exc: Exception) -> None:
+        # If the task fails, and we have retries left, set the task to retrying.
+        if not await self.handle_retry(exc):
+            # If the task has no retries left, or the retry condition is not met, set the task to failed.
+            context = TaskRunContext.get()
+            state = await exception_to_failed_state(
+                exc,
+                message="Task run encountered an exception",
+                result_factory=getattr(context, "result_factory", None),
+            )
+            await self.set_state(state)
+
+    async def handle_crash(self, exc: BaseException) -> None:
+        state = await exception_to_crashed_state(exc)
+        self.logger.error(f"Crash detected! {state.message}")
+        self.logger.debug("Crash details:", exc_info=exc)
+        await self.set_state(state, force=True)
+
+    async def create_task_run(self, client: PrefectClient) -> TaskRun:
+        flow_run_ctx = FlowRunContext.get()
+        try:
+            task_run_name = _resolve_custom_task_run_name(self.task, self.parameters)
+        except TypeError:
+            task_run_name = None
+
+        # prep input tracking
+        task_inputs = {
+            k: await collect_task_run_inputs(v) for k, v in self.parameters.items()
+        }
+
+        # anticipate nested runs
+        task_run_ctx = TaskRunContext.get()
+        if task_run_ctx:
+            task_inputs["wait_for"] = [TaskRunResult(id=task_run_ctx.task_run.id)]
+
+        #        if wait_for:
+        #            task_inputs["wait_for"] = await collect_task_run_inputs(wait_for)
+
+        if flow_run_ctx:
+            dynamic_key = _dynamic_key_for_task_run(
+                context=flow_run_ctx, task=self.task
+            )
+        else:
+            dynamic_key = uuid4().hex
+        task_run = await client.create_task_run(
+            task=self.task,
+            name=task_run_name,
+            flow_run_id=(
+                getattr(flow_run_ctx.flow_run, "id", None)
+                if flow_run_ctx and flow_run_ctx.flow_run
+                else None
+            ),
+            dynamic_key=dynamic_key,
+            state=Pending(),
+            task_inputs=task_inputs,
+        )
+        return task_run
+
+    @asynccontextmanager
+    async def enter_run_context(self, client: Optional[PrefectClient] = None):
+        if client is None:
+            client = await self.get_client()
+
+        self.task_run = await client.read_task_run(self.task_run.id)
+
+        with TaskRunContext(
+            task=self.task,
+            log_prints=self.task.log_prints or False,
+            task_run=self.task_run,
+            parameters=self.parameters,
+            result_factory=await ResultFactory.from_autonomous_task(self.task),
```

Comment:
hmm 
```suggestion
            result_factory=await ResultFactory.from_autonomous_task(self.task), # TODO why not `from_task`
```

Reply:
The result interface confuses me so I wasn't sure what methods were safe to use here -- because the task engine is designed to not be conditional on autonomous vs. non-autonomous tasks, I thought this method call was safer?  Looking at the results code confuses me more, because it looks like `from_autonomous_task` is actually called when the task is... not autonomous?